File: app/src/main/java/com/foreverrafs/numericals/fragments/conversions/FragmentAllInOne.java
package com.foreverrafs.numericals.fragments.conversions;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;

import com.foreverrafs.core.Numericals;
import com.foreverrafs.numericals.R;
import com.foreverrafs.numericals.utils.Utilities;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;

import butterknife.BindView;
import butterknife.ButterKnife;


/**
 * Created by Aziz Rafsanjani on 11/4/2017.
 */

public class FragmentAllInOne extends Fragment implements View.OnClickListener, TextWatcher {

    @BindView(R.id.layout_answer_area)
    ConstraintLayout layoutAnswerArea;


    @BindView(R.id.btnCalculate)
    Button btnCalculate;

    @BindView(R.id.btnBack)
    Button btnBack;

    @BindView(R.id.til_user_input)
    TextInputLayout tilUserInput;

    @BindView(R.id.tvAnswerBinary)
    TextView tvBinary;

    @BindView(R.id.tvAnswerOctal)
    TextView tvOctal;

    @BindView(R.id.tvAnswerHexadecimal)
    TextView tvHexadecimal;


    private View rootView;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_conversions_allinone, container, false);

        ButterKnife.bind(this, rootView);
        initControls();
        return rootView;
    }

    private void initControls() {

        EditText etInput = tilUserInput.getEditText();

        etInput.setOnKeyListener((view, i, keyEvent) -> {
            //take the error message away
            tilUserInput.setErrorEnabled(false);

            if (keyEvent.getKeyCode() == KeyEvent.KEYCODE_ENTER) {
                onCalculate();
                return true;
            }
            return false;
        });

        etInput.addTextChangedListener(this);

        btnBack.setOnClickListener(this);
        btnCalculate.setOnClickListener(this);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btnBack:
                break;

            case R.id.btnCalculate:
                onCalculate();
                break;
        }
    }

    private void onCalculate() {
        TextInputEditText etInput = rootView.findViewById(R.id.text_user_input);
        String binary, octal, hexadecimal;

        String decimal = etInput.getText().toString();
        if (decimal.isEmpty()) {
            tilUserInput.setError("Input field is empty");
            return;
        }

        try {
            double decLong = Double.parseDouble(decimal);

            if (decLong <= 0) {
                tilUserInput.setError("Number should be greater than 0");
                return;
            }

            binary = Numericals.decimalToBinary(decLong);
            octal = Numericals.decimalToOctal(decimal);
            hexadecimal = Numericals.decimalToHexadecimal(decimal);

            tvBinary.setText(binary);
            tvOctal.setText(octal);
            tvHexadecimal.setText(hexadecimal);

            Utilities.animateAnswer(layoutAnswerArea,
                    rootView, Utilities.DisplayMode.SHOW);


        } catch (NumberFormatException ex) {
            Log.e(Utilities.LOG_TAG, "cannot parse " + decimal + " to an integer value");
        } catch (Exception ex) {
            Log.e(Utilities.LOG_TAG, ex.getMessage());
        }
    }


    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void afterTextChanged(Editable editable) {
        if (editable.length() == 0) {
            Utilities.animateAnswer(layoutAnswerArea,
                    rootView, Utilities.DisplayMode.HIDE);
        }
    }
}

File: app/src/main/java/com/foreverrafs/numericals/ui/conversions/FragmentAllInOne.kt
package com.foreverrafs.numericals.ui.conversions

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.fragment.app.Fragment
import com.foreverrafs.core.Numericals
import com.foreverrafs.numericals.activities.MainActivity
import com.foreverrafs.numericals.databinding.FragmentConversionsAllinoneBinding
import com.foreverrafs.numericals.utils.Utilities
import com.google.android.material.textfield.TextInputLayout
import timber.log.Timber

/** Created by Aziz Rafsanjani on 11/4/2017. */
class FragmentAllInOne : Fragment(), TextWatcher {
    private lateinit var layoutAnswerArea: ConstraintLayout

    lateinit var btnCalculate: Button

    private lateinit var btnBack: Button

    private lateinit var tilUserInput: TextInputLayout

    private lateinit var tvBinary: TextView

    private lateinit var tvOctal: TextView

    private lateinit var tvHexadecimal: TextView

    lateinit var binding: FragmentConversionsAllinoneBinding;

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentConversionsAllinoneBinding.inflate(inflater)
        initControls()
        return binding.root
    }

    private fun initControls() {
        tvHexadecimal = binding.tvAnswerHexadecimal
        tvOctal = binding.tvAnswerOctal
        tvBinary = binding.tvAnswerBinary
        tilUserInput = binding.tilUserInput
        btnBack = binding.btnBackToMainMenu
        layoutAnswerArea = binding.layoutAnswerArea;
        btnCalculate = binding.btnCalculate;


        binding.tilUserInput.setOnKeyListener { _: View?, i: Int, keyEvent: KeyEvent ->
            //take the error message away
            tilUserInput.isErrorEnabled = false
            if (keyEvent.keyCode == KeyEvent.KEYCODE_ENTER) {
                onCalculate()
                return@setOnKeyListener true
            }
            false
        }
        binding.tilUserInput.editText?.addTextChangedListener(this)

        btnBack.setOnClickListener { onBackToMainMenu() }

        btnCalculate.setOnClickListener { onCalculate() }
    }

    private fun onBackToMainMenu() {
        (activity as MainActivity?)!!.goToMainMenu()
    }

    fun onCalculate() {
        val binary: String
        val octal: String
        val hexadecimal: String
        val decimal = tilUserInput.editText?.text.toString()

        if (decimal.isEmpty()) {
            tilUserInput.error = "Input field is empty"
            return
        }
        try {
            val decLong = decimal.toDouble()
            if (decLong <= 0) {
                tilUserInput.error = "Number should be greater than 0"
                return
            }
            binary = Numericals.decimalToBinary(decLong)
            octal = Numericals.decimalToOctal(decimal)
            hexadecimal = Numericals.decimalToHexadecimal(decimal)
            tvBinary.text = binary
            tvOctal.text = octal
            tvHexadecimal.text = hexadecimal
            Utilities.animateAnswer(
                layoutAnswerArea,
                binding.root, Utilities.DisplayMode.SHOW
            )
        } catch (ex: NumberFormatException) {
            Timber.e("cannot parse $decimal to an integer value : $ex")
        } catch (ex: Exception) {
            Timber.e(ex)
        }
    }

    override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
    override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
    override fun afterTextChanged(editable: Editable) {
        if (editable.isEmpty()) {
            Utilities.animateAnswer(
                layoutAnswerArea,
                binding.root, Utilities.DisplayMode.HIDE
            )
        }
    }
}